### PR TITLE
Add MCopy to Evm 

### PR DESCRIPTION
**Pull request by Demuirgos**
## Changes

- Adds initial implementation of Eip:5656 
- Adds MCOPY opcode to EVM

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 1]
```diff
@@ -1773,6 +1773,32 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
...
+                        Metrics.MCopyOpcode++;
+
+                        stack.PopUInt256(out UInt256 dstPosition);
+                        stack.PopUInt256(out UInt256 srcPosition);
+                        stack.PopUInt256(out UInt256 length);
+
+                        long baseGasCost = GasCostOf.VeryLow * (((long)length + 31) / 32);
+                        if (!UpdateGas(baseGasCost, ref gasAvailable)) goto OutOfGas;
+
+                        if (!UpdateMemoryCost(vmState, ref gasAvailable, in srcPosition, length)) goto OutOfGas;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1794]:
```
words_copied = (length + 31) // 32
g_verylow    = 3
g_copy       = 3 * words_copied + memory_expansion_cost
gas_cost     = g_verylow + g_copy
```

I think you missed last `g_verylow` in your gas calculations?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 2]
```diff
@@ -1773,6 +1773,32 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
...
+                        Span<byte> loadedData = vmState.Memory.LoadSpan(in srcPosition, length);
+                        if (_txTracer.IsTracingInstructions) _txTracer.ReportMemoryChange(srcPosition, loadedData);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1796]: Memory is changing on load?

Comment (Demuirgos) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1796]: I was thinking of removing these and keeping on the loading line, and then saw that MLOAD does them as well, so I kept them

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 3]
```diff
@@ -1545,6 +1545,31 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
...
+                        stack.PopUInt256(out UInt256 length);
+
+                        long baseGasCost = GasCostOf.VeryLow + GasCostOf.VeryLow * (((long)length + 31) / 32);
+                        if (!UpdateGas(baseGasCost, ref gasAvailable)
+                            || !UpdateMemoryCost(vmState, ref gasAvailable, UInt256.Max(srcPosition, dstPosition), length)) goto OutOfGas;
+
+                        Span<byte> loadedData = vmState.Memory.LoadSpan(in srcPosition, length);
+                        if (_txTracer.IsTracingInstructions) _txTracer.ReportMemoryChange(srcPosition, loadedData);
+
+                        vmState.Memory.Save(in dstPosition, loadedData);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1570]: This should handle overlapping memory just fine, right?

Comment (Demuirgos) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1570]: I wrote a test case for this scenario and it passes, if you have lore complex scenarios to test please do tell

Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1570]: It calls `Span<byte>.CopyTo` which calls into `Buffer.Memmove` which has a check and handling for overlapping buffers in the runtime https://github.com/dotnet/runtime/blob/8770047887d05d106699badd27520a7b260a99be/src/libraries/System.Private.CoreLib/src/System/Buffer.cs#L131-L137

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1570]: Yeah I thought so, just double-checking :)

**Review by MarekM25 — CHANGES_REQUESTED** [review]
@Demuirgos it won't be possible to configure this EIP in chainspec

**Review by smartprogrammer93 — APPROVED** [review]

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 4]
```diff
@@ -1545,6 +1545,31 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
...
+                            goto InvalidInstruction;
+                        }
+
+                        Metrics.MCopyOpcode++;
+
+                        stack.PopUInt256(out UInt256 dstPosition);
+                        stack.PopUInt256(out UInt256 srcPosition);
+                        stack.PopUInt256(out UInt256 length);
+
+                        long baseGasCost = GasCostOf.VeryLow + GasCostOf.VeryLow * (((long)length + 31) / 32);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1563]: What will happen if the length will be >long.MaxValue? possible scenario?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 5]
```diff
@@ -1545,6 +1545,31 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                         vmState.Memory.SaveByte(in memPosition, data);
                         if (_txTracer.IsTracingInstructions) _txTracer.ReportMemoryChange((long)memPosition, data);
 
+                        break;
+                    }
+                case Instruction.MCOPY:
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1550]: btw do you know if we have Ethereum state tests for MCOPY or not yet?

Comment (Demuirgos) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs]: there is this [PR](https://github.com/ethereum/tests/pull/1229/files) that was shared today

File: src/Nethermind/Nethermind.Evm.Test/VirtualMachineTests.cs [hunk 6]
```diff
@@ -427,6 +430,75 @@ public void Tload()
             Assert.That(receipt.GasSpent, Is.EqualTo(GasCostOf.Transaction + GasCostOf.VeryLow * 1 + GasCostOf.TLoad), "gas");
         }
 
+        /// <summary>
+        /// MCOPY gas cost check
+        /// </summary>
+        [Test]
+        public void MCopy()
```
[the remark this answers is not in the document]
Comment (Demuirgos) [reply src/Nethermind/Nethermind.Evm.Test/VirtualMachineTests.cs:437]: I think we had all scenarios :
- copy to same location
- copy to a location without overlapping destination with source 
- copy to a location with overlapping destination with source

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs (hunk 5 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs]: Can we run them? :)

Comment (Demuirgos) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs]: will try to do it asap

**Review by MarekM25 — APPROVED** [review]

**Pull request merged**